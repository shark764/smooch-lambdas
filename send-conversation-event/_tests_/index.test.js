const axios = require('axios');

jest.mock('axios');

const event = {
  params: {
    'tenant-id': '66d83870-30df-4a3b-8801-59edff162034',
    'interaction-id': '66d83870-30df-4a3b-8801-59edff162040',
  },
  identity: {
    'user-id': '667802d8-2260-436c-958a-2ee0f71f73f0',
    name: 'name',
  },
  body: {
    event: 'conversation-read',
  },
};

const mockConversationActivity = jest.fn()
  .mockImplementation(() => ({
    promise: () => ({}),
  }));

const mockSmoochCore = jest.fn(() => ({
  appUsers: {
    conversationActivity: mockConversationActivity,
  },
}));

const mockGetSecretValue = jest.fn(() => {})
  .mockImplementation(() => ({
    promise: () => ({
      SecretString: JSON.stringify({
        '5e31c81640a22c000f5d7f28-id': 'id',
        '5e31c81640a22c000f5d7f28-secret': 'secret',
      }),
    }),
  }));

axios.mockImplementation(() => ({
  data: {
    appId: '5e31c81640a22c000f5d7f28',
    userId: '5e31c81640a22c000f5d7f70',
    method: 'get',
    url: 'https://us-east-1-dev-edge.domain/v1/tenants/66d83870-30df-4a3b-8801-59edff162034/interactions/66d83870-30df-4a3b-8801-59edff162040/metadata',
  },
}));

jest.mock('aws-sdk', () => ({
  SecretsManager: jest.fn().mockImplementation(() => ({
    getSecretValue: mockGetSecretValue,
  })),
}));

jest.mock('smooch-core', () => mockSmoochCore);

const { handler } = require('../index');

describe('send-conversation-event', () => {
  describe('Everthing is successful', () => {
    afterEach(() => {
      jest.clearAllMocks();
    });
    it("when userEvent is 'typing-start'", async () => {
      const typingStartEvent = {
        ...event,
        body: { event: 'typing-start' },
      };
      await handler(typingStartEvent);
      expect(mockConversationActivity.mock.calls[0][0].activityProps).toEqual({
        role: 'appMaker',
        type: 'typing:start',
      });
    });

    it("when userEvent is 'typing-stop'", async () => {
      const typingStopEvent = {
        ...event,
        body: { event: 'typing-stop' },
      };
      await handler(typingStopEvent);
      expect(mockConversationActivity.mock.calls[0][0].activityProps).toEqual({
        role: 'appMaker',
        type: 'typing:stop',
      });
    });

    it('sends back status 200 when the code runs without any error', async () => {
      const result = await handler(event);
      expect(result).toMatchSnapshot();
    });

    describe('Walkthrough', () => {
      beforeEach(async () => {
        mockGetSecretValue.mockImplementationOnce(() => ({
          promise: () => ({
            SecretString: JSON.stringify({
              '5e31c81640a22c000f5d7f28-id': 'id',
              '5e31c81640a22c000f5d7f28-secret': 'secret',
            }),
          }),
        }));
        mockGetSecretValue.mockImplementationOnce(() => ({
          promise: () => ({
            SecretString: JSON.stringify({
              username: 'username',
              password: 'paasword',
            }),
          }),
        }));
        await handler(event);
      });
      it('passes in the correct arguments to secretClient.getSecretValue() to get digital channels credentials', async () => {
        expect(mockGetSecretValue.mock.calls[0]).toEqual(expect.arrayContaining([{
          SecretId: 'us-east-1-dev-smooch-app',
        }]));
        expect(mockGetSecretValue.mock.calls[0]).toMatchSnapshot();
      });

      it('passes in the correct arguments to secretClient.getSecretValue() to get cx credentials', async () => {
        expect(mockGetSecretValue.mock.calls[1]).toEqual(expect.arrayContaining([{
          SecretId: 'us-east-1-dev-smooch-cx',
        }]));
        expect(mockGetSecretValue.mock.calls[1]).toMatchSnapshot();
      });

      it('passes in the correct arguments to axios to get the interaction metadata', async () => {
        expect(axios.mock.calls).toMatchSnapshot();
      });

      it('passes in the correct arguments to SmoochCore', async () => {
        expect(mockSmoochCore.mock.calls).toMatchSnapshot();
      });

      it('passes in the correct arguments to smooch.appUsers.conversationActivity()', async () => {
        expect(mockConversationActivity.mock.calls).toMatchSnapshot();
      });
    });
  });

  it('sends back status 500 when there is an error retrieving digital channels credentials', async () => {
    mockGetSecretValue.mockRejectedValueOnce(new Error());
    const result = await handler(event);
    expect(result).toMatchSnapshot();
  });

  it('sends back status 500 when there is an error retrieving cx credentials', async () => {
    mockGetSecretValue.mockImplementationOnce(() => ({
      promise: () => ({
        SecretString: JSON.stringify({
          '5e31c81640a22c000f5d7c55-id': 'id',
          '5e31c81640a22c000f5d7c55-secert': 'secret',
        }),
      }),
    }));
    mockGetSecretValue.mockRejectedValueOnce(new Error());
    const result = await handler(event);
    expect(result).toMatchSnapshot();
  });

  it('sends back status 500 when there is a error retrieving interaction metadata', async () => {
    axios.mockRejectedValueOnce(new Error());
    const result = await handler(event);
    expect(result).toMatchSnapshot();
  });

  it('sends back status 400 when the provided event is not supported', async () => {
    event.body.event = 'event';
    const result = await handler(event);
    expect(result).toMatchSnapshot();
  });

  it('sends back status 500 when there is an error sending conversation activity', async () => {
    event.body.event = 'conversation-read';
    mockConversationActivity.mockRejectedValueOnce(new Error());
    const result = await handler(event);
    expect(result).toMatchSnapshot();
  });

  it('sends back status 500 when there is a error validating digital channels credentials', async () => {
    mockSmoochCore.mock.calls = null;
    const result = await handler(event);
    expect(result).toMatchSnapshot();
  });
});
