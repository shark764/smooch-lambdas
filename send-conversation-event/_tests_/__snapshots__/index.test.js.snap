// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`send-conversation-event Everthing is successful Walkthrough passes in the correct arguments to SmoochCore 1`] = `
Array [
  Array [
    Object {
      "keyId": "id",
      "scope": "app",
      "secret": "secret",
      "serviceUrl": "mock-smooch-api-url",
    },
  ],
]
`;

exports[`send-conversation-event Everthing is successful Walkthrough passes in the correct arguments to axios to get the interaction metadata 1`] = `
Array [
  Array [
    Object {
      "auth": Object {
        "password": "paasword",
        "username": "username",
      },
      "method": "get",
      "url": "https://us-east-1-dev-edge.domain/v1/tenants/66d83870-30df-4a3b-8801-59edff162034/interactions/66d83870-30df-4a3b-8801-59edff162040/metadata",
    },
  ],
]
`;

exports[`send-conversation-event Everthing is successful Walkthrough passes in the correct arguments to secretClient.getSecretValue() to get cx credentials 1`] = `
Array [
  Object {
    "SecretId": "us-east-1-dev-smooch-cx",
  },
]
`;

exports[`send-conversation-event Everthing is successful Walkthrough passes in the correct arguments to secretClient.getSecretValue() to get digital channels credentials 1`] = `
Array [
  Object {
    "SecretId": "us-east-1-dev-smooch-app",
  },
]
`;

exports[`send-conversation-event Everthing is successful Walkthrough passes in the correct arguments to smooch.appUsers.conversationActivity() 1`] = `
Array [
  Array [
    Object {
      "activityProps": Object {
        "role": "appMaker",
        "type": "conversation:read",
      },
      "appId": "5e31c81640a22c000f5d7f28",
      "userId": "5e31c81640a22c000f5d7f70",
    },
  ],
]
`;

exports[`send-conversation-event Everthing is successful sends back status 200 when the code runs without any error 1`] = `
Object {
  "body": Object {
    "event": "conversation-read",
    "resourceId": "667802d8-2260-436c-958a-2ee0f71f73f0",
  },
  "status": 200,
}
`;

exports[`send-conversation-event sends back status 400 when the provided event is not supported 1`] = `
Object {
  "body": Object {
    "message": "the provided event is not supported",
  },
  "status": 400,
}
`;

exports[`send-conversation-event sends back status 500 when there is a error retrieving interaction metadata 1`] = `
Object {
  "body": Object {
    "message": "An error occurred retrieving the interaction metadata",
  },
  "status": 500,
}
`;

exports[`send-conversation-event sends back status 500 when there is a error validating digital channels credentials 1`] = `
Object {
  "body": Object {
    "message": "An Error has occurred trying to validate digital channels credentials",
  },
  "status": 500,
}
`;

exports[`send-conversation-event sends back status 500 when there is an error retrieving cx credentials 1`] = `
Object {
  "body": Object {
    "message": "An Error has occurred trying to retrieve cx credentials",
  },
  "status": 500,
}
`;

exports[`send-conversation-event sends back status 500 when there is an error retrieving digital channels credentials 1`] = `
Object {
  "body": Object {
    "message": "An Error has occurred trying to retrieve digital channels credentials",
  },
  "status": 500,
}
`;

exports[`send-conversation-event sends back status 500 when there is an error sending conversation activity 1`] = `
Object {
  "body": Object {
    "message": "An error occurred trying to send conversation activity",
  },
  "status": 500,
}
`;
